internal/controllers/job_event_reconciler.go

```diff
@@ -0,0 +1,232 @@
+
+			return ctrl.Result{}, r.removeFinalizer(ctx, &pod)
+		} else {
+			return ctrl.Result{}, fmt.Errorf("error while getting Module %s: %v", moduleNSN, err)
+		}
+	}
+
+	eventAnnotations := map[string]string{
+		"kernel-version": kernelVersion,
+		"pod-name":       pod.Name,
+	}
+
+	if _, ok := pod.GetAnnotations()[createdAnnotationKey]; !ok {
+		patchFrom := client.MergeFrom(pod.DeepCopy())
+
+		meta.SetAnnotation(&pod, createdAnnotationKey, "")
+
+		if err := r.client.Patch(ctx, &pod, patchFrom); err != nil {
+			return ctrl.Result{}, fmt.Errorf("could not patch Pod %s: %v", req.NamespacedName, err)
+		}
+
+		ann := maps.Clone(eventAnnotations)
+		ann["pod-creation-timestamp"] = pod.CreationTimestamp.String()
+
+		r.recorder.AnnotatedEventf(
+			&mod,
+			ann,
+			v1.EventTypeNormal,
+			je.ReasonCreated(),
+			"%s created for kernel %s",
+			je,
+			kernelVersion,
+		)
+	}
+
+	if controllerutil.ContainsFinalizer(&pod, constants.JobEventFinalizer) {
+		var eventType, fmtString, reason string
+
+		switch pod.Status.Phase {
+		case v1.PodFailed:
+			eventType = v1.EventTypeWarning
+			fmtString = "%s job failed for kernel %s"
+			reason = je.ReasonFailed()
+		case v1.PodSucceeded:
+			eventType = v1.EventTypeNormal
+			fmtString = "%s job succeeded for kernel %s"
+			reason = je.ReasonSucceeded()
+		default:
+			// still running, nothing to do
+			return ctrl.Result{}, nil
+		}
+
+		if err := r.removeFinalizer(ctx, &pod); err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not patch pod %s: %v", req.NamespacedName, err)
+		}
+
+		r.recorder.AnnotatedEventf(
+			&mod,
+			eventAnnotations,
+			eventType,
+			reason,
+			fmtString,
+			req.NamespacedName,
+			kernelVersion,
+		)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *JobEventReconciler) SetupWithManager(mgr manager.Manager) error {
+	p, err := JobEventPodPredicate()
+	if err != nil {
+		return fmt.Errorf("could not build the Pod predicate: %v", err)
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(
+			&v1.Pod{},
+			builder.WithPredicates(p),
+		).
+		Named(JobEventReconcilerName).
+		Complete(r)
+}
+
+func (r *JobEventReconciler) removeFinalizer(ctx context.Context, pod *v1.Pod) error {
+	if controllerutil.ContainsFinalizer(pod, constants.JobEventFinalizer) {
+		patchFrom := client.MergeFrom(pod.DeepCopy())
+
+		controllerutil.RemoveFinalizer(pod, constants.JobEventFinalizer)
+
+		if err := r.client.Patch(ctx, pod, patchFrom); err != nil {
+			return fmt.Errorf("patch failed: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func JobEventPodPredicate() (predicate.Predicate, error) {
+	isJobPod, err := filter.JobPod()
+	if err != nil {
+		return nil, fmt.Errorf("error building the predicate for job Pod: %v", err)
+	}
+
+	hasJobEventFinalizer := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return controllerutil.ContainsFinalizer(o, constants.JobEventFinalizer)
+	})
+
+	p := predicate.And(
+		isJobPod,
+		predicate.Or(
+			hasJobEventFinalizer,
+			predicate.Not(
```

Comment:
Why do we need predicate for not having Annotation? Why not just check job finalizer existence?